In [10]:
import time

import ray

##### Example 1

In [15]:
ray.shutdown()
ray.init()

2023-03-22 13:59:03,576	INFO worker.py:1553 -- Started a local Ray instance.


RayContext(dashboard_url='', python_version='3.8.16', ray_version='2.3.0', ray_commit='cf7a56b4b0b648c324722df7c99c168e92ff0b45', address_info={'node_ip_address': '127.0.0.1', 'raylet_ip_address': '127.0.0.1', 'redis_address': None, 'object_store_address': '/tmp/ray/session_2023-03-22_13-59-00_476951_4432/sockets/plasma_store', 'raylet_socket_name': '/tmp/ray/session_2023-03-22_13-59-00_476951_4432/sockets/raylet', 'webui_url': '', 'session_dir': '/tmp/ray/session_2023-03-22_13-59-00_476951_4432', 'metrics_export_port': 62913, 'gcs_address': '127.0.0.1:58453', 'address': '127.0.0.1:58453', 'dashboard_agent_listen_port': 52365, 'node_id': 'b4af10c0e574ba1dc951faf271d922e776987cbf6b3fd8de75e43a71'})

In [16]:
ray.cluster_resources()

{'object_store_memory': 2147483648.0,
 'CPU': 8.0,
 'memory': 12102409831.0,
 'node:127.0.0.1': 1.0}

In [17]:
database = [
    "Learning", "Ray",
    "Flexible", "Distributed", "Python", "for", "Machine", "Learning"
]

In [18]:
def retrieve(item):
    time.sleep(item / 10.)
    return item, database[item]

In [19]:
type(retrieve)

function

In [20]:
%timeit data = [retrieve(item) for item in range(8)] 

2.82 s ± 3.04 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


Look! It's took **2.82 s** to run the code above. Now execute the operation above in parallel across multiple nodes in the cluster

In [21]:
import ray

In [22]:
@ray.remote
def retrieve_task(item):
    return retrieve(item)

In [23]:
%timeit data = [retrieve_task.remote(item) for item in range(8)] 

501 µs ± 31.7 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


In [27]:
@ray.remote
class DataTracker:
    def __init__(self):
        self._counts = 0
    
    def increment(self): self._counts += 1
    
    def counts(self): return self._counts